flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java

```diff
@@ -74,88 +76,233 @@
+			if (usedNames.contains(currentPattern.getName())) {
+				throw new MalformedPatternException(
+					"Duplicate pattern name: " + currentPattern.getName() + ". " +
+					"Pattern names must be unique.");
+			}
+		}
+
+		@SuppressWarnings("unchecked")
+		private State<T> createStartState(State<T> sinkState) {
+			final State<T> beginningState;
+
+			checkPatternNameUniqueness();
+
+			if (currentPattern.getQuantifier().isLooping()) {
+				final State<T> loopingState;
+				if (currentPattern.getQuantifier().isAtLeastOne()) {
+					loopingState = new State<>(currentPattern.getName(), State.StateType.Normal);
+					beginningState = createFirstMandatoryStateOfLoop(loopingState, State.StateType.Start);
+					states.add(loopingState);
 				} else {
-					currentState = new State<>(currentPattern.getName(), State.StateType.Normal);
-					states.put(currentState.getName(), currentState);
+					loopingState = new State<>(currentPattern.getName(), State.StateType.Start);
+					beginningState = loopingState;
 				}
-
-				currentState.addStateTransition(new StateTransition<T>(
-					StateTransitionAction.TAKE,
-					succeedingState,
-					(FilterFunction<T>) succeedingPattern.getFilterFunction()));
-
-				if (succeedingPattern instanceof FollowedByPattern) {
-					// the followed by pattern entails a reflexive ignore transition
-					currentState.addStateTransition(new StateTransition<T>(
-						StateTransitionAction.IGNORE,
-						currentState,
-						null
-					));
+				convertToLooping(sinkState, loopingState);
+			} else if (currentPattern.getQuantifier() == Quantifier.TIMES) {
+				if (currentPattern.getTimes() > 1) {
+					final State<T> timesState = new State<>(currentPattern.getName(), State.StateType.Normal);
+					states.add(timesState);
+					sinkState = convertToTimesState(sinkState, timesState, currentPattern.getTimes() - 1);
 				}
+				beginningState = new State<>(currentPattern.getName(), State.StateType.Start);
+				beginningState.addTake(sinkState, (FilterFunction<T>) currentPattern.getFilterFunction());
+			} else {
+				beginningState = new State<>(currentPattern.getName(), State.StateType.Start);
+				beginningState.addTake(sinkState, (FilterFunction<T>) currentPattern.getFilterFunction());
 			}
 
-			// add the beginning state
-			final State<T> beginningState;
+			states.add(beginningState);
+			usedNames.add(beginningState.getName());
+
+			return beginningState;
+		}
+
+		private State<T> convertToTimesState(State<T> sinkState, State<T> sourceState, int times) {
+			convertToSingletonState(sinkState, sourceState);
+			for (int i = 0; i < times - 1; i++) {
+				sinkState = sourceState;
+				sourceState = new State<>(currentPattern.getName(), State.StateType.Normal);
+				states.add(sourceState);
+				convertToSingletonState(sinkState, sourceState);
+			}
+			return sourceState;
+		}
+
+		@SuppressWarnings("unchecked")
+		private void convertToSingletonState(
+			final State<T> sinkState,
```

Comment:
I've done it the other way round as to handle optional in one place. But I think your version is easier to analyze.